### DISCUSSION TITLE
Support for other PCF8574 like devices (TCA6408A) how to contribute

### DISCUSSION THREAD

**Discussion by plewka** [arendst/Tasmota#14026] [Ideas]
I dublicated and modified the driver of that I2C-IO-Expander to support my hardware.
It worked immediately (only output). Seems I understood how to at least
do the most important things ;-)
I added two new defines (USE_ and MAX_) and added it to the feature 0x200.
 
For me this device (in result) behaves equivalent enough to the PCF8574 for 
higher level. It could be probed to divide from a real PCF8574 and
there are 16bit derivates on the other hand.
It's basicly not as powerfull as the stuff from microchip (MCP320xx).

In my opinion it doesn't make too much sense to mix these devices 
(PCF8574 vs. PCF8574) on one bus.
What's the right idea to contribute? As a faked PCF it uses the global 
structures, configuration webpage etc. of the PCF8574 without overhead.

Is it prefered/better to keep the code fully separate, do alternative code 
for one driver ID or to use conditional compile and mix it all in one?

https://github.com/plewka/Tasmota/blob/development/tasmota/xdrv_28_tca6408a.ino

**Comment by barbudor** [comment]
Not sure to get the value of this device vs the PCF8574
We already have PC8574 and MCP230xx as I/O extender and their driver are complicated enough to maintain.
Why would we add a 3rd one ?

**Reply by plewka** [reply]
You're right if you're free to decide which device is used which is not always the case.
This is why I would probably simply treat it as a PCF8574 if there is no benefit.
Basically you don't have to care for anything else but modified read/write to/from registers 
and the direction init in the code
The Reset might be quite usefull to immediately disable things at the IOs, but that's beyond software.

**Reply by barbudor** [reply]
I've not been through the datasheet nor your code.
Do you mean that the chip is basically compatible with PCF (same register map) and will work as a PCF if you use the PCF driver ?

If this is the case, the best could be not to create a new driver but to add a #define such as #USE_TCA6408A_EXTENSIONS that would enable only the additional features supported by the TCA.
So with #define USE_PCF8574 you get a driver that works with legacy PCF8574 and TCA6408 in compatibility mode and by adding #USE_TCA6408A_EXTENSIONS, you would enable the advanced features.

You can also create a totally new driver of course but not with number 28.
You need to get the next available XDRV (87) and the next available I2C (65)
You also need to change any reference to PCF8574 into TCA such as variable names including constants like HTTP_FORM_I2C_PCF8574_1

**Comment by plewka** [comment]
No, the PCF doesn't use real addressing of registers. I2C-read gets the inputs, write sets the outputs, that's it. 
TCA has registers for input, output and data direction - plus (quite useless) inverting, 
only targeting the inputs.

Beyond software it offers a reset pin and can follow a system reset immediately. The PCF does poor voltage level adaption (pull-up) and there can be issues with input leakage. 
Electrically the outputs behave different. The PCF has logical states "H/0" as output states while the 
TCA has got full PushPull "1/0" but no build in pullups on the other hand (sometimes usefull for the inputs).

The driver code makes it possible to degrade it as a PCF for normal use case since it masks out the inputs,
which is equivalent to setting a data direction. You basically don't loose functionality...ok, you could create
a tristate-output...very rare use ;-)
 
It's a political question of a newbie what matches best. There are many of these I2C GPIOs I agree and 
I don't want to "waste" a driver id or add duplicate code...
The TCA offers two possible addresses (which are part of the PCFs blocks) and probably it's possible 
to detect it is not a PCF because of the registers. 

I can try do a minimum change version of the PCF driver. This would target the device 
search, setting of the data direction and wrapping of I2C read+write by using #ifdefined-masks. What do you think? 
Plus a TREAT_TCA_AS_PCF #define which steals both addresses from PCF like already done for the MCP...

**Reply by barbudor** [reply]
ok, so if the registers map are different, different chips, different drivers
so you have to create as a new driver with new DRV id and new I2C id
It's the responsibility of the user who compiles this driver into its build to not-compile or disable (with I2Cdriver) the drivers for conflicting addresses.
There are other cases in the I2CDEVICES tables of conflicting addresses and user must know what they are doing

Note : Independently of my advices on how to best make the driver, I cannot commit that Theo will accept to merge a driver for a 3rd IO expander, whatever are its advantages. So you may end-up with a driver that will stay only for you (which is fine too if you have the hardware for it)

**Reply by plewka** [reply]
Ok, thanks a lot!
I just have had a look at our "parts used in past" database.
The driver would even give direct support for PCA6408A and PCA9557 while PCA9574 
offers additional features like bushold and programmable pullup/down (other register mapping).
I don't have them on ESP hardware, though.

Maybe its possible to enable more of these TI and Philips/NXP devices without 
too much if then else like with the BMEs. I'll try a more generic approach. 
For me in low level code is easy such changes but if it comes to higher level I'll probably
have questions.

For first I need to add DS2482-x and DS2485.pdf 
I would be very thankful to hear an advice/discussion on how to optionally "replace" the bitbanging OneWire
to the I2C way. I tried to ask, but it went a helpfull, but other way. Can this be done with an I2C driver
or is there an architectural conflict?
https://github.com/arendst/Tasmota/discussions/13932
